redis/credentials.py

```diff
@@ -0,0 +1,77 @@
+from typing import Callable, Optional, Union
+
+
+class CredentialProvider:
```

Comment:
I'm not sure why this needs to be a class at all – couldn't a provider just be a function `get_credentials(*, username, password)`...

Reply:
Different credential providers will need different arguments to be passed, so we can save within the credential provider object the required args/kwargs.

Reply:
You're right: a credential provider could indeed just be a `get_credentials()` function with no knowledge of _any_ arguments. 😉 

```python
def get_very_specific_credential_provider(some_auth_url):
    def provider():
        # ... do some API call..?
        return ("foo", "bar")
    return provider

c = Client(
    credential_provider=get_very_specific_credential_provider("https://secret..."),
)
```

Reply:
Yup, that would do the job. :)
@chayim WDYT?

Reply:
There's no reason that during the instantiation of all Redis connection objects a function can't be passed in, as opposed to the class being discussed. However, I think the current class implementation is more readable for **the community that uses this**. I'm partial for erring towards readability, and the clear separation it provides.

Reply:
@chayim @barshaul Well, this class looks like it's become quite complicated and hard-to-follow.

It has an username/password that may not be used at all, args and kwargs that may not be used at all.

If you want a class, then why not just 

```
class CredentialProvider:
    def get_credentials(self) -> Union[Tuple[str], Tuple[str, str]]:
        raise NotImplementedError("get_credentials must be implemented")


class UsernamePasswordCredentialProvider(CredentialProvider):
    def __init__(self, username: Optional[str]=None, password: Optional[str]=None):
        self.username = username
        self.password = password

    def get_credentials(self):
        if self.username:
            return (self.username, self._password)
        return (self.password,)
```

and leave the rest (e.g. implementing a specific provider with a specific `get_credentials`) to the users?

As an aside – should `get_credentials` maybe also support being async? :)

Reply:
With the changes now done to `Connection`, I really do believe that the implementation I suggest above would be the better fit, and the current `CredentialProvider` is _way_ too generic. I sincerely ask you to revise it.

Reply:
```get_credentials``` should definitely support an async all, to keep this aligned.  

On the class inputs side - I don't think it's unreasonable to pass in an optional ```CredentialProvider``` on a per redis-connection type. I see classes as more useful given state needs, depending on the provider. Maybe instead, ```CredentialProvider``` shouldn't require any specified arguments, instead be a descendent of an ABC that accepts ```**kwargs```. This could address the same need, and reduce the complexity, but yes - be perhaps generic enough (though perhaps something shared).

@akx WDYT needlessly generic? Having a way to do this for systems that are completely different makes it problematic.

Reply:
@chayim The "needlessly generic" comment was outdated by https://github.com/redis/redis-py/pull/2261/commits/abe613773b070754f54e7e797d3a7af022673700, which removed the "username + password + supplier + args + kwargs" mode.